src/ucs/sys/topo.c

```diff
@@ -19,9 +52,138 @@ ucs_sys_bus_id_t ucs_sys_bus_id_unknown = { .domain   = 0xffff,
+    kh_init_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+    kh_init_inplace(sys_dev_to_bus, &ucs_topo_sys_dev_to_bus_hash);
+
+    ucs_topo_initialized = 1;
+}
+
+void ucs_topo_cleanup()
+{
+    ucs_status_t status;
+    ucs_sys_device_t sys_dev;
+    ucs_sys_bus_id_t bus_id;
+
+    if (!ucs_topo_initialized) return;
+
+    kh_foreach_value(&ucs_topo_bus_to_sys_dev_hash, sys_dev, {
+        ucs_debug("sys_device_t %u not released", sys_dev);
+    })
+
+    kh_destroy_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_bus_to_sys_dev_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    kh_foreach_value(&ucs_topo_sys_dev_to_bus_hash, bus_id, {
+        ucs_debug("bus id %ld not released",
+                   ucs_topo_get_bus_id_bit_repr(&bus_id));
+    })
+
+    kh_destroy_inplace(sys_dev_to_bus, &ucs_topo_sys_dev_to_bus_hash);
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_sys_dev_to_bus_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    ucs_topo_initialized = 0;
+}
+
+static int ucs_topo_compare_bus_id(const ucs_sys_bus_id_t *bus_id1, 
+                                   const ucs_sys_bus_id_t *bus_id2)
+{
+    if (bus_id1->domain == bus_id2->domain &&
+        bus_id1->bus == bus_id2->bus &&
+        bus_id1->slot == bus_id2->slot &&
+        bus_id1->function == bus_id2->function) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
 ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
                                             ucs_sys_device_t *sys_dev)
 {
+    khiter_t hash_it;
+    //ucs_status_t status;
+    ucs_topo_kh_put_t kh_put_status;
+
+    if (ucs_topo_compare_bus_id(bus_id, &ucs_sys_bus_id_unknown)) {
+        *sys_dev = UINT_MAX;
+        ucs_debug("found unknown device index %u for bus id %ld",
+                  *sys_dev, ucs_topo_get_bus_id_bit_repr(bus_id));
+        return UCS_OK;
+    }
+    
+    *sys_dev = UINT_MAX;
+    ucs_debug("find device index for bus id %ld", ucs_topo_get_bus_id_bit_repr(bus_id));
+
+    ucs_recursive_spin_lock(&ucs_topo_bus_to_sys_dev_lock);
+    hash_it = kh_put(bus_to_sys_dev/*name*/, 
+                     &ucs_topo_bus_to_sys_dev_hash /*pointer to hashmap*/,
+                     ucs_topo_get_bus_id_bit_repr(bus_id)/*key*/,
+                     &kh_put_status);
+
+    switch (kh_put_status) {
+        case UCS_KH_PUT_KEY_PRESENT:
+            *sys_dev = kh_value(&ucs_topo_bus_to_sys_dev_hash, hash_it);
+            ucs_debug("bus id %ld exists. sys_dev = %u",
+                      ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
+            break;
+        case UCS_KH_PUT_BUCKET_EMPTY:
+            *sys_dev = kh_size(&ucs_topo_sys_dev_to_bus_hash);
```

Comment:
maybe use `ptr_array` instead? It returns an unsigned index..
Also can't we just keep sys_dev id in the bus id?

Reply:
@brminich Haven't addressed this yet but will look ucs/datastruct/ptr_array and will reply back here.

Reply:
@brminich can you clarify if the suggestion is to use `ptr_array` instead of khash for sys_device_t index -> bus_id lookup? If yes, I agree that we don't actually need a hashmap. 

How about using a simple array `static ucs_sys_bus_id_t ucs_topo_sys_dev_to_bus_arr[UCS_MAX_SYS_DEVICES];` in `ucs/sys/topo.c` instead though? After all, we end up populating device indices 0 to N after all and I don't think the upper limit grows beyond say 512 or 1024 for a given system. LMK if using a simple array is ok.

Also, we can cache sys_dev in bus_id and not have a separate structure for sys_dev to bus_id lookup but a) we need to change ucs topo API for this. b) we may need to populate bus_id  members each time we want to query distance.

Reply:
@Akshay-Venkatesh, simple array sounds good as well.
can you please elaborate more on b)?

Reply:
@brminich by b) I mean that if we decide to populate `bus_id->sys_dev` in `find_device_by_bus_id`, then we may need to hold references to `populate bus_id->sys_dev` before calling `ucs_topo_get_distance` to refer to sys_device_t indices which are passed to it. The other issue is that we would have to change the API of `find_device_by_bus_id` because it uses a constant parameter bus_id and it's non-intuitive that `find_device_by_bus_id` should modify the contents of bus_id. But I do agree now that holding on to the value of system device in bus_id is probably quite convenient.